### PR TITLE
oauth2-server: added certificate availability check

### PR DESCRIPTION
**Pull request by VasilMitovClouway**
Id_token no longer created in IssueNewTokenActivity if there are no certificates configured.
related #45

### CODE REVIEW

**Comment by VasilMitovClouway** [comment]
Cannot add labels or assign reviewers.

File: oauth2-server/src/test/java/com/clouway/oauth2/IssueNewTokenForClientTest.java [hunk 1]
```diff
@@ -94,20 +94,13 @@ public void tokenCannotBeIssued() throws Exception {
 
       oneOf(identityFinder).findIdentity(with(any(String.class)), with(any(GrantType.class)), with(any(DateTime.class)));
       will(returnValue(Optional.of(identity)));
-
```
Comment (mgenov) [inline oauth2-server/src/test/java/com/clouway/oauth2/IssueNewTokenForClientTest.java]: I'm not able to see test for the new behaviour.

**Review by mgenov — CHANGES_REQUESTED** [review]

File: oauth2-server/src/test/java/com/clouway/oauth2/IssueNewTokenForClientTest.java (hunk 1 above)
Comment (VasilMitovClouway) [reply oauth2-server/src/test/java/com/clouway/oauth2/IssueNewTokenForClientTest.java]:
```java
@Test
  public void publicCertificatesNotConfigured() throws Exception {
    IssueNewTokenActivity controller = new IssueNewTokenActivity(tokens, identityFinder, clientAuthorizationRepository, keyStore);
    final Client client = aNewClient().build();
    final DateTime anyTime = new DateTime();
    final Identity identity = aNewIdentity().withId("::user_id::").build();

    context.checking(new Expectations() {{
      oneOf(clientAuthorizationRepository).findAuthorization(with(any(Client.class)), with(any(String.class)));
      will(returnValue(Optional.of(new Authorization("", "", "::user_id::", "::auth_code::", Collections.<String>emptySet(), Collections.singleton("::redirect_uri::")))));

      oneOf(tokens).issueToken(GrantType.AUTHORIZATION_CODE, client, identity, Collections.<String>emptySet(), anyTime);
      will(returnValue(new TokenResponse(true, aNewToken().withValue("::token::").build(), "::refresh token::")));

      oneOf(identityFinder).findIdentity(with(any(String.class)), with(any(GrantType.class)), with(any(DateTime.class)));
      will(returnValue(Optional.of(identity)));

      oneOf(keyStore).publicCertificates();
      will(returnValue(Collections.emptyMap()));
    }});

    Response response = controller.execute(client, new ParamRequest(ImmutableMap.of("code", "::auth_code::", "redirect_uri", "::redirect_uri::")), anyTime);
    String body = new RsPrint(response).printBody();

    assertThat(body, containsString("::token::"));
  }
```

The thing is that this test is an absolute copy of the happy path with the difference that public certs return empty map and I didnt think it was informative enought to be added.

Comment (mgenov) [reply oauth2-server/src/test/java/com/clouway/oauth2/IssueNewTokenForClientTest.java]: It indicates a smell in the code as this check need to be performed here which actually need to be considered.

Comment (mgenov) [reply oauth2-server/src/test/java/com/clouway/oauth2/IssueNewTokenForClientTest.java]: Keep in mind that signing need to be performed by private keys and verification to be performed with public keys, which means that name of the method is misleading and requires some fixes.

Comment (mgenov) [reply oauth2-server/src/test/java/com/clouway/oauth2/IssueNewTokenForClientTest.java]: The second comment is not going to be addressed here unless you see it as reasonable to be here.

**Comment by VasilMitovClouway** [comment]
Regarding the availability check, we had a discussion about the ClientKeyStore.certificates() to return a Certificate object. Here is an example implementation of the Certificate Object and a helper class CertificateParcer

```java
public class Certificate {
  public final String certId;
  public final String privatePem;
  public final String publicPem;

  public Certificate(String certId, String privatePem, String publicPem) {
    this.certId = certId;
    this.privatePem = privatePem;
    this.publicPem = publicPem;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Certificate that = (Certificate) o;
    return Objects.equal(certId, that.certId) &&
            Objects.equal(privatePem, that.privatePem) &&
            Objects.equal(publicPem, that.publicPem);
  }

  @Override
  public int hashCode() {
    return Objects.hashCode(certId, privatePem, publicPem);
  }
}
```
```java
public class CertificateParcer {

  public static PrivateKey getPrivateKey(Certificate certificate) {
    Block block = getPemBlock(certificate.privatePem);
    return (PrivateKey) getKey("private",block);
  }

  public static PublicKey getPublicKey(Certificate certificate) {
    Block block = getPemBlock(certificate.privatePem);
    return (PublicKey) getKey("public",block);
  }

  public static Pem.Block getPrivatePem(Certificate certificate) {
    return getPemBlock(certificate.privatePem);
  }

  public static Pem.Block getPublicPem(Certificate certificate) {
    return getPemBlock(certificate.publicPem);
  }

  private static Key getKey(String type,Block block){
    try {
      PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(block.getBytes());
      KeyFactory kf = KeyFactory.getInstance("RSA");
      switch (type){
        case "private":return kf.generatePrivate(keySpec);
        case "public":return kf.generatePublic(keySpec);
      }
    } catch (NoSuchAlgorithmException|InvalidKeySpecException e) {
      e.printStackTrace();
    }
    return null;
  }

  private static Block getPemBlock(String pemString) {
    Pem pem = new Pem();
    try {
      Pem.Block block = pem.parse(new ByteArrayInputStream(pemString.getBytes()));
      return block;
    } catch (IOException e) {
      e.printStackTrace();
    }
    return null;
  }
}
```

**Comment by mgenov** [comment]
Maybe it would be better client apps to pass PrivateKey and PublicKey objects directly instead of PEM. This will let them decide in what form to be persisted. 

About this design, it's not visible how this code is going to be used by the IssueNewTokenActivity. The problem there is that it has many dependencies which causes tests to blow.

**Comment by mgenov** [comment]
About the Certificate Class: 

Keep in mind that Java already has such classes:
https://docs.oracle.com/javase/7/docs/api/java/security/cert/Certificate.html
https://docs.oracle.com/javase/7/docs/api/java/security/cert/X509Certificate.html

and there is a difference between private/public key and certificate (certificate adds additional information about the owner of the certificate). It would be better library to support certificates instead of simple keys.

Take a look at: 
https://www.googleapis.com/oauth2/v1/certs

As you can see it serves CERTIFICATES, not keys.

**Comment by VasilMitovClouway** [comment]
I think now the tests are more readable. Moved the keystore.privateCertificates() a bit lower, because it is not needed to be done when Oauth.error is returned shortening a few tests and it is now easier to read the tests when certificates are not returned after #55 .

File: oauth2-server/src/test/java/com/clouway/oauth2/IssueNewTokenForClientTest.java [hunk 2]
```diff
@@ -65,6 +65,27 @@ public void happyPath() throws IOException {
   }
 
   @Test
+  public void idTokenNotReturnedWhenCertificatesNotImplemented() throws Exception {
```
Comment (mgenov) [inline oauth2-server/src/test/java/com/clouway/oauth2/IssueNewTokenForClientTest.java]: noPrivateCertsAreConfigured

File: oauth2-server/src/main/java/com/clouway/oauth2/client/CertificatesNotConfiguredException.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package com.clouway.oauth2.client;
+
+/**
+ * @author Vasil Mitov <v.mitov.clouway@gmail.com>
+ */
+public class CertificatesNotConfiguredException extends RuntimeException {
```
Comment (mgenov) [inline oauth2-server/src/main/java/com/clouway/oauth2/client/CertificatesNotConfiguredException.java]: This exception is not used.

File: oauth2-server/src/main/java/com/clouway/oauth2/IssueNewTokenActivity.java [hunk 4]
```diff
@@ -43,29 +43,31 @@
...
-            .setIssuedAt(instant.asDate())
-            .setExpiration(new Date(accessToken.expirationTimestamp()))
-            .claim("identity", identityAsJson(identity).toString())
-            .signWith(SignatureAlgorithm.RS256, parsePem(key))
-            .compact();
-
+    Map<String, Block> certificates = keyStore.privateCertificates();
+    String idToken = "";
+
+    if (available(certificates)) {
```
Comment (mgenov) [inline oauth2-server/src/main/java/com/clouway/oauth2/IssueNewTokenActivity.java]: I'm not sure that null check need to be added. It's bad practices methods to return null but we cannot relay on correct behaviour from client apps. 

it's more readable if it's inline:

```java
if (certificates != null && !certificates.isEmpty()) {

}
```

**Review by mgenov — CHANGES_REQUESTED** [review]

File: oauth2-server/src/test/java/com/clouway/oauth2/IssueNewTokenForClientTest.java (hunk 2 above)
Comment (VasilMitovClouway) [reply oauth2-server/src/test/java/com/clouway/oauth2/IssueNewTokenForClientTest.java]: Tried to be as specific as I can in the description. Name is now changed

File: oauth2-server/src/main/java/com/clouway/oauth2/client/CertificatesNotConfiguredException.java (hunk 3 above)
Comment (VasilMitovClouway) [reply oauth2-server/src/main/java/com/clouway/oauth2/client/CertificatesNotConfiguredException.java]: Removed

File: oauth2-server/src/main/java/com/clouway/oauth2/IssueNewTokenActivity.java (hunk 4 above)
Comment (VasilMitovClouway) [reply oauth2-server/src/main/java/com/clouway/oauth2/IssueNewTokenActivity.java]: Removed private method and moved the checks inline. We have tested this if block, and it fixes the bug that occurred last week

**Review by mgenov — APPROVED** [review]
LGTM !

**Pull request merged**